### ISSUE TITLE
Binary Classification - Last layer

### ISSUE THREAD

**Issue by res97** [vlfeat/matconvnet#55]
I am using the same network as the one used for imagenet for binary classification. I did change the flag to binary. However, at the testing stage, I still get 1000 outputs for the last layer, even though only first two classes are dominant, and other 998 classes have extremely low values, and for practical purposes can be ignored. What is the issue here really?  I'm wondering that can you please help me so that I get only two output values.

**Comment by lenck** [comment]
Hi, at first, the files in folder 'examples' is sort of example usage of the library, so it is generally just one possible use-case of the library, usually for existing networks. So for your own task, you probably need to modify those files to your particular needs. This is also a reason why changing the `errorType` to `binary` in `examples/cnn_train.m` basically wont have any effect on your training because it influences only the way how error is evaluated.
So, for binary classification on imagenet you should also modify the architecture of the network so it would have only two outputs, e.g. learning new FC8 layer with two outputs. And then you will also have to modify the `updateError` in `examples/cnn_train.m` so it won't compute top5 error...

**Comment by res97** [comment]
Thank you for your answer. However, it seems like that there is no facility for adding fc8 layer in the package. Should we write our own fc8 layer or is there something convenient in the package that we can use.

**Comment by lenck** [comment]
It's easy - just directly change the network structure, so in your case basically setting up to have 2 filters and biases in the last layer and initialising them with for example random numbers. The whole network structure with all its weights is in that single matlab structure, no additional files or definitions are needed....

Example how the network is created is in e.g. [cnn_imagenet.m](https://github.com/vlfeat/matconvnet/blob/master/examples/cnn_imagenet.m) on lines 98-213. So, I guess in your case you just take the pre-trained imagenet network, find the FC8 layer with filters of size [1,1,4096,1000] in `net.layers` cell array and reinitialise them with smaller bank of filters (however it must be 'single' data type). Same for the biases...

**Comment by res97** [comment]
Thank you, Karel, for your response. There are few more things, if you can possibly help with these, that would be great:

1. In the last layer before softmax, that is 'conv' layer, I changed filters and biases to 2. Is this the only thing to be changed? Also, it is 'conv' layer, NOT fc8 layer. Is that an issue?

net.layers{end+1} = struct('type', 'conv', ...
                           'filters', 0.01/scal * randn(1,1,4096,2,'single'), ... % changed 1000 to 2
                           'biases', zeros(1, 2, 'single'), ...        % changed 1000 to 2
                           'stride', 1, ...
                           'pad', 0, ...
                           'filtersLearningRate', 1, ...
                           'biasesLearningRate', 2, ...
                           'filtersWeightDecay', 1, ...
                           'biasesWeightDecay', 0) ;

2. For binary classification, using this approach, I did not get good results. I used 100 images each for 2 categories for training. Should I increase number of training images, or should I change some parameters or architecture? Any suggestions.

**Comment by lenck** [comment]
Yes, this seems fine :)
And the 'fc8' is just sort of a name of this layer and the type should be 'conv'...

And the issue with training - I think it's too little training examples. But why don't you try e.g. linear SVM on tom of e.g. FC7? That is usually the simplest thing to do and it usually works quite well...

**Comment by res97** [comment]
Thanks again, Karel. I would like to take a bit more of your time:) I have couple questions:

1. If I use SVM, then it will be two stages of training. In the stage 1, I train the cnn model using softmaxloss. In the second stage, I train SVM using the same training set as was used for cnn. Right?

2. I guess there is no svm in Matconvnet. I will have to use liblinear or another svm after extracting features from the last 'conv' layer. And while extracting features, I should make all the features as one-dimensional array. For example, [1,1,4096, 2] will give me 8192 features. Right?

**Comment by bazilas** [comment]
1. Right. You will remove the softmaxloss layer and use the output of the last fully connected layer as descriptor for each sample.

2. Try VLFeat (also from Vedaldi) or LIBSVM. They both work with Matlab. The softmaxloss layer is the one of  [1,1,4096, 2]. You won't use this for the SVM. Instead, you will use the previous one (output 4096 values) as descriptor.

cheers,
Vasilis

**Comment by lenck** [comment]
You welcome :)
One article where SVM is being used is [RCNN](http://arxiv.org/abs/1311.2524), however they use it for detection, so they have a large amount of negative samples for fine-tuning.
But if you have just few samples, I would think that the fine-tuning would not bring that much improvement and if you would train with them too much, you can start to overfit...
And there is no SVM in matconvnet - but there is one in VLFeat (the reason for all those cryptic `vl_` names), the 'mother project' of matconvnet. Or there is also liblinear, which does the same, but you know, I must tell you about vlfeat first ;)
And the features to use - I would recommend to test outputs from FC6 and FC7, which will work better. And  `[1,1,4096, 2]` looks more like size of the filters which you have on top, your feature vector from FC7 would be `[1,1,4096, N]` where N is number of images, so your weight vector `w` will be of size 4096. What you typically do for the FCx features before feeding them into svm training is just:
```
x = permute(res(fcx_buf_idx).x, [3 4 1 2])
```
Which will create you a vector of size `[4096, N]`.
I hope it helps :)

**Comment by res97** [comment]
"your feature vector from FC7 would be [1,1,4096, N] where N is number of images,"

I'm a bit confused here. Why N should be number of images? I think I was extracting features from one image only.

**Comment by lenck** [comment]
You can process arbitrary number of images at once (they are aligned along the 4th dimension of the input and output array) and you are limited only by the memory...

**Comment by swangb** [comment]
Hi,
I also meet some problems when doing binary classification. I start with the pretrained imagenet-caffe-ref model and use two-class image data (face and background)  880(440 440) or 1760(440 1320) to fine-tune the model. I did the modification to the fc7 and also errorType as described above. I also change the learning ratio to be 10 times of origin for fc7. After the code run a few epoch, the object result decrease reasonably, however the error rate always converges to the ratio of data: for 880 data the error rate is 50% and for 1760 data the error is 25%. Is this the bug of the binary errorType evaluation?
Thank you!

**Comment by Devangin9** [comment]
Hi,

According to http://www.robots.ox.ac.uk/~vgg/practicals/cnn/index.html#part-33-learning-with-gradient-descent, the weights are updated according to both the learning rate and momentum. So, shouldn't the momentum be 1 and learning rate be 0 for the fixed layers and then momentum be around 0.9 and learning rate be low e.g. 0.00001 inorder to do fine-tuning?

**Comment by lenck** [comment]
Hi, that is a local learning rate and momentum per each layer. If you look to [cnn_train](https://github.com/vlfeat/matconvnet/blob/master/examples/cnn_train.m) it uses the global learning rate from the `opts` structure and multiplies it with the layers' options.
So, in this way you can easily e.g. freeze a layer by setting its learning rate to zero...

**Comment by swangb** [comment]
Hi,
I think in the fin-tuning, you should not freeze any layer. Recognizing Image Style paper re-initialize the fully connect layer and give it larger local learning rate than the convolutional layer. But the global learning rate should be smaller for fine-tuning, from example use 0.0001 as start point for imagenet fine-tune.

**Comment by Almonfrey** [comment]
@lenck I'm not finding the place where I can set the momentum locally, as you said. I tried this:
-------------------------------------------------------------------------------------------------
net.layers{32} = struct('type', 'conv', 'name', 'fc6', ...
'weights', {{0.005*randn(2,4,512,2048, 'single'), zeros(1,2048,'single')}}, ...
'learningRate', [10 20], ...
'weightDecay', [1 0], ...
'momentum', {1 1}, ...
'stride', [1 1], ...
'pad', [0 0 0 0]) ;
-------------------------------------------------------------------

but I got the following error:

Error using strcmp
Too many input arguments.

Error in vl_simplenn_tidy (line 40)
  if strcmp(layer.type, 'custom')

Error in cnn_train (line 56)
net = vl_simplenn_tidy(net); % fill in some eventually missing values

Error in transferLearning (line 96)
[net, info] = trainfn(net, imdb, getBatch(opts), ...

**Comment by Almonfrey** [comment]
I found the problem... Apparently I just had to do    ->       'momentum', {{1 1}}, ... 
But, when you initialize direct in the net variable, just do this -> net.layers{i}.momentum = {1 1};

However, in the cnn_train.m there is a variable called state.layers{l}.momentum{j} that remains equal zero even when I initialize the momentum...

**Comment by h612** [comment]
@lenk i cant figure out how to use svm layer before layer fc7, for instance. 1. I cant imagine the addlayer syntax for SVM layer?. 2. Also the output of svm (SVR to be precise) how will it pass the output to fc7? 3.How does that go through backpropagation?

**Comment by h612** [comment]
@lenck
